core/src/main/java/com/linecorp/armeria/server/cors/CorsService.java

```diff
@@ -108,6 +109,58 @@ protected HttpObject filter(HttpObject obj) {
         };
     }
 
+    /**
+     * This is a non CORS specification feature which enables the setting of preflight
+     * response headers that might be required by intermediaries.
+     *
+     * @param headers the {@link HttpHeaders} to which the preflight headers should be added.
+     */
+    void setPreflightHeaders(final CorsPolicy policy, final HttpHeaders headers) {
+        requireNonNull(policy, "policy");
+        for (Map.Entry<AsciiString, String> entry : policy.preflightResponseHeaders()) {
+            headers.add(entry.getKey(), entry.getValue());
+        }
+    }
+
+    void setCorsAllowCredentials(final CorsPolicy policy, final HttpHeaders headers) {
+        requireNonNull(policy, "policy");
+        if (policy.isCredentialsAllowed() &&
+            !headers.get(HttpHeaderNames.ACCESS_CONTROL_ALLOW_ORIGIN).equals(CorsService.ANY_ORIGIN)) {
+            headers.set(HttpHeaderNames.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+        }
+    }
+
+    void setCorsExposeHeaders(final CorsPolicy policy, final HttpHeaders headers) {
+        requireNonNull(policy, "policy");
+        if (policy.exposedHeaders().isEmpty()) {
+            return;
+        }
+
+        headers.set(HttpHeaderNames.ACCESS_CONTROL_EXPOSE_HEADERS, HEADER_JOINER.join(policy.exposedHeaders()));
```

Comment:
How about keeping the precalculated value of the headers set by `setCors*()` in `CorsPolicy`? Then, it would make sense to move the setter methods to `CorsPolicy`.